tests/providers/amazon/aws/operators/test_redshift.py

```diff
@@ -0,0 +1,57 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import unittest
+from unittest import mock
+
+from airflow.models.dag import DAG
+from airflow.providers.amazon.aws.operators.redshift import RedshiftOperator
+from airflow.utils import timezone
+
+DEFAULT_DATE = timezone.datetime(2015, 1, 1)
+DEFAULT_DATE_ISO = DEFAULT_DATE.isoformat()
+DEFAULT_DATE_DS = DEFAULT_DATE_ISO[:10]
+TEST_DAG_ID = 'unit_test_dag'
+
+
+class TestRedshiftOperator(unittest.TestCase):
+    def setUp(self):
+        super().setUp()
+        args = {'owner': 'airflow', 'start_date': DEFAULT_DATE}
+        dag = DAG(TEST_DAG_ID, default_args=args)
+        self.dag = dag
+
+    @mock.patch("airflow.providers.amazon.aws.operators.redshift.RedshiftOperator.get_hook")
+    def test_redshift_operator(self, mock_get_hook):
+        sql = """
+        CREATE TABLE IF NOT EXISTS test_airflow (
+            dummy VARCHAR(50)
+        );
+        """
+        operator = RedshiftOperator(task_id='redshift_operator', sql=sql, dag=self.dag)
+        operator.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
+
+    @mock.patch("airflow.providers.amazon.aws.operators.redshift.RedshiftOperator.get_hook")
+    def test_redshift_operator_test_multi(self, mock_get_hook):
+        sql = [
+            "CREATE TABLE IF NOT EXISTS test_airflow (dummy VARCHAR(50))",
+            "TRUNCATE TABLE test_airflow",
+            "INSERT INTO test_airflow VALUES ('X')",
+        ]
+        operator = RedshiftOperator(task_id='redshift_operator_test_multi', sql=sql, dag=self.dag)
+        operator.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
```

Comment:
```suggestion
import unittest
from unittest import mock
from unittest.mock import MagicMock

from airflow.providers.amazon.aws.operators.redshift import RedshiftOperator


class TestRedshiftOperator(unittest.TestCase):
    @mock.patch("airflow.providers.amazon.aws.operators.redshift.RedshiftOperator.get_hook")
    def test_redshift_operator(self, mock_get_hook):
        hook = MagicMock()
        mock_run = hook.run
        mock_get_hook.return_value = hook
        sql = MagicMock()
        operator = RedshiftOperator(task_id='test', sql=sql)
        operator.execute(None)
        mock_run.assert_called_once_with(
            sql,
            autocommit=True,
            parameters=None,
        )
```

Ok so why have I suggested this rewriting...

If you look at your test, the behavior of the operator is not verified.  `get_hook` will return a mock, and nothing that done to the mock will cause an error, and the calls are not verified.

So for example, if I swap out the list of string like so it still passes:
```python
    @mock.patch("airflow.providers.amazon.aws.operators.redshift.RedshiftOperator.get_hook")
    def test_redshift_operator_test_multi(self, mock_get_hook):
        def fun():
            print('hello')

        sql = fun
        operator = RedshiftOperator(task_id='redshift_operator_test_multi', sql=sql, dag=self.dag)
        operator.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
```

So the there is no difference between your test for str and the one for list[str]; and neither of them really verify the behavior of the operator.

Further, you don't really need to verify that hook.run takes list of str or str -- that is behavior of DBAPIHook and presumably verified elsewhere.  We just care that hook.run is called with the values passed to the operator.

So that's what I have tried to do here.

And I called execute directly instead of run because it's simpler and means we can chop some of the boilerplate.

You could improve upon my suggestion by parameterizing it such that it verifies that autocommit and parameters are forward to hook.run as appropriate

Reply:
thank you for providing the re-write and very thorough explanation. i've addressed this in 7205a8cd7d0d05ed70dc7e5f87bf70d0227f0184, and added some parameterization for autocommit and parameters